### PR TITLE
statistics: revise loaded for column

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What problem does this PR solve?

Issue Number:  ref https://github.com/pingcap/tidb/issues/34052

Problem Summary:

### What is changed and how it works?

revise loaded for column in order to keep compatible if the statistics of the column got evicted.

### Check List

Tests 

- [] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: statistics/handle/handle.go [hunk 1]
```diff
@@ -831,7 +836,7 @@ func (h *Handle) columnStatsFromStorage(reader *statsReader, row chunk.Row, tabl
 		// 4. loadAll is false.
 		notNeedLoad := h.Lease() > 0 &&
 			!isHandle &&
-			(col == nil || !col.IsLoaded() && col.LastUpdateVersion < histVer) &&
+			(col == nil || !col.IsHistogramLoaded() && col.LastUpdateVersion < histVer) &&
```
Comment (xuyifangreeneyes) [inline statistics/handle/handle.go:839]: If all the samples are in TopN and histogram is empty, `col.Histogram.notNullCount()` is zero, i.e., `col.IsHistogramLoaded()` is false. Even if all the column stats have been loaded, we will enter `notNeedLoad` branch in that case. Maybe `IsLoaded` or `IsNecessaryLoaded` is more suitable here.

File: statistics/histogram.go [hunk 2]
```diff
@@ -1069,6 +1070,37 @@ func (c *Column) IsLoaded() bool {
 	return c.Loaded || c.notNullCount() > 0
 }
 
+// IsNecessaryLoaded indicates whether the necessary statistics is loaded.
+// If `IsLoaded` returns true, we will directly return due to they are directly loaded from outer storage and
+// keep the tests still correct.
+// If `IsLoaded` returns false, we will check whether histogram and topn are both loaded as the statistics need
+// at least one of them to do optimize.
+func (c *Column) IsNecessaryLoaded() bool {
```
Comment (xuyifangreeneyes) [inline statistics/histogram.go:1078]: There seems to be little difference between `IsNecessaryLoaded` and `IsLoaded`. Can we just keep one?

File: statistics/handle/handle.go (hunk 1 above)
Comment (Yisaer) [reply statistics/handle/handle.go]: updated.

**Review by xuyifangreeneyes — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**